enterprise/coderd/license/metricscollector.go

```diff
@@ -0,0 +1,77 @@
+package license
+
+import (
+	"sync/atomic"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/coder/coder/v2/codersdk"
+)
+
+type MetricsCollector struct {
+	Entitlements atomic.Pointer[codersdk.Entitlements]
+
+	activeUsersGauge      prometheus.Gauge
+	limitUsersGauge       prometheus.Gauge
+	userLimitEnabledGauge prometheus.Gauge
+}
+
+var _ prometheus.Collector = new(MetricsCollector)
+
+func NewMetricsCollector() *MetricsCollector {
+	return &MetricsCollector{
+		activeUsersGauge: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "coderd",
+			Subsystem: "license",
+			Name:      "active_users",
+			Help:      `The number of active users.`,
+		}),
+		limitUsersGauge: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "coderd",
+			Subsystem: "license",
+			Name:      "limit_users",
+			Help:      "The user seats limit based on the active Coder license.",
+		}),
+		userLimitEnabledGauge: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "coderd",
+			Subsystem: "license",
+			Name:      "user_limit_enabled",
+			Help:      "Returns 1 if the current license enforces the user limit.",
+		}),
+	}
+}
+
+func (mc *MetricsCollector) Describe(descCh chan<- *prometheus.Desc) {
+	descCh <- mc.activeUsersGauge.Desc()
+	descCh <- mc.limitUsersGauge.Desc()
+	descCh <- mc.userLimitEnabledGauge.Desc()
+}
+
+func (mc *MetricsCollector) Collect(metricsCh chan<- prometheus.Metric) {
+	entitlements := mc.Entitlements.Load()
+	if entitlements == nil || entitlements.Features == nil {
+		return
+	}
+
+	userLimitEntitlement, ok := entitlements.Features[codersdk.FeatureUserLimit]
+	if !ok {
+		return
+	}
+
+	var enabled float64
+	if userLimitEntitlement.Enabled {
+		enabled = 1
+	}
+	mc.userLimitEnabledGauge.Set(enabled)
+	metricsCh <- mc.userLimitEnabledGauge
```

Comment:
The canonical way to do this is to create constant metrics on the fly for the instantaneous values, rather than a long-lived gauge.

You create package-level, constant `Desc` for each with https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#NewDesc

Then, on `Collect` send a constant metric with the current value:

https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#MustNewConstMetric

--- 

Passing mutable gauges could allow them to be changed by a concurrent call to `Collect` while being processed, which could result in an inconsistent "snapshot" of the metrics.  Probably doesn't matter in this case, since active users changes slowly, and the limit changes rarely, but in more dynamic data you could get inconsistencies that cause bad downstream effects like false alarms.  But, still a good idea to do it the canonical way.

Reply:
Done. I replaced all metrics with const substitutes.